File: Chowdren/base/objects/colorizer.cpp
#include "colorizer.h"

ColorizerObject::ColorizerObject(int x, int y, int id)
: FrameObject(x, y, id), r(1.0f), g(1.0f), b(1.0f)
{

}

inline float round_color(float v)
{
    return int(v * 256.0f) / 256.0f;
}

void ColorizerObject::set_red(float v)
{
    r = round_color(v);
}

void ColorizerObject::set_green(float v)
{
    g = round_color(v);
}

void ColorizerObject::set_blue(float v)
{
    b = round_color(v);
}

File: Chowdren/chowdren/writers/extensions/Colorizer.py
from chowdren.writers.objects import ObjectWriter
from chowdren.common import get_animation_name, to_c, make_color
from chowdren.writers.events import (ComparisonWriter, ActionMethodWriter,
    ConditionMethodWriter, ExpressionMethodWriter, make_table,
    EmptyAction)
from mmfparser.data.font import LogFont


class ColorizerObject(ObjectWriter):
    class_name = 'ColorizerObject'
    filename = 'colorizer'
    use_alterables = True

    def write_init(self, writer):
        pass

actions = make_table(ActionMethodWriter, {
    29 : EmptyAction, # load mask from overlay/surface
    32 : EmptyAction, # change mask mode
    0 : 'set_red', # set red out of old red
    4 : 'set_green', # set green out of old green,
    8 : 'set_blue' # set blue out of old blue
})

conditions = make_table(ConditionMethodWriter, {
})

expressions = make_table(ExpressionMethodWriter, {
    2 : '.r', # red out of red,
    6 : '.g', # green out of old green
    10 : '.b' # blue out of old blue
})

def get_object():
    return ColorizerObject
